chainbase/src/main/java/org/tron/core/db2/common/TxCacheDB.java

```diff
@@ -224,6 +242,152 @@ public void close() {
+      return true;
+    }).exceptionally(e -> {
+      bloomFilters[0] = BloomFilter.create(Funnels.byteArrayFunnel(),
+          MAX_BLOCK_SIZE * TRANSACTION_COUNT);
+      bloomFilters[1] = BloomFilter.create(Funnels.byteArrayFunnel(),
+          MAX_BLOCK_SIZE * TRANSACTION_COUNT);
+      logger.warn("recovery bloomFilters failed. {}", e.getMessage());
+      logger.info("rollback to previous mode.");
+      return false;
+    }).join();
+  }
+
+  private boolean recovery(int index, Path file) {
+    try (InputStream in = new BufferedInputStream(Files.newInputStream(file,
+        StandardOpenOption.READ, StandardOpenOption.DELETE_ON_CLOSE))) {
+      logger.info("recovery bloomFilter[{}] from file.", index);
+      long start = System.currentTimeMillis();
+      bloomFilters[index] = BloomFilter.readFrom(in, Funnels.byteArrayFunnel());
+      logger.info("recovery bloomFilter[{}] from file done,filter: {}, filter-fpp: {}, cost {} ms.",
+          index, bloomFilters[index].approximateElementCount(), bloomFilters[index].expectedFpp(),
+          System.currentTimeMillis() - start);
+      return true;
+    } catch (IOException e) {
+      logger.warn("recovery bloomFilter[{}] failed.", index, e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void dump() {
+    logger.info("dump bloomFilters start.");
+    Path file0 = Paths.get(CommonParameter.getInstance().getOutputDirectory(),
+        "bloomFilters_0");
+    Path file1 = Paths.get(CommonParameter.getInstance().getOutputDirectory(),
+        "bloomFilters_1");
+    Path txCacheProperties = Paths.get(CommonParameter.getInstance().getOutputDirectory(),
+        "txCache.properties");
+    CompletableFuture<Void> task0 = CompletableFuture.runAsync(() -> dump(0, file0));
+    CompletableFuture<Void> task1 = CompletableFuture.runAsync(() -> dump(1, file1));
+    CompletableFuture.allOf(task0, task1).thenRun(() -> {
+      logger.info("filterStartBlock: {}, currentBlockNum: {}, currentFilterIndex: {}",
+          filterStartBlock, currentBlockNum, currentFilterIndex);
+      Map<String, String> properties = new HashMap<>();
+      properties.put("filterStartBlock", String.valueOf(filterStartBlock));
+      properties.put("currentBlockNum", String.valueOf(currentBlockNum));
+      properties.put("currentFilterIndex", String.valueOf(currentFilterIndex));
+      writeProperties(txCacheProperties, properties);
+      logger.info("dump bloomFilters done.");
+
+    }).exceptionally(e -> {
+      logger.warn("dump bloomFilters to file failed. {}", e.getMessage());
+      return null;
+    }).join();
+  }
+
+  private void dump(int index, Path file) {
+    try {
+      Files.deleteIfExists(file);
+    } catch (IOException e) {
+      logger.warn("dump bloomFilters[{}] failed.", index, e);
+      throw new RuntimeException(e);
+    }
+    try (OutputStream out = new BufferedOutputStream(Files.newOutputStream(file,
+        StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE))) {
```

Comment:
> Given that OS cache flushing to the disk is not immediate, is it possible that the property file flushed successfully but partly of bloomFilters is still in the OS cache(then corrupt when OS crashed)?

Yeah, it's gonna happen. But it doesn't seem to be a problem, it will rollback to previous mode.

Reply:
> Add another option StandardOpenOption.SYNC? Or write all data into one file and make sure the metadata at the end also can achieve the atomic, but this will introduce complexity

This PR is a helper feature. `StandardOpenOption.SYNC` is good , But it may slow down the closing procedure if you're so concerned about closing.

Reply:
How long does `StandardOpenOption.SYNC` take?

Reply:
300ms -> 1.2 s for ssd

Reply:
700ms -> +15s for non-ssd, `StandardOpenOption.SYNC` is not a good choice!